### ISSUE TITLE
Issue with type inference on class member of a Literal type 

### ISSUE THREAD

**Issue by SandyChapman** [microsoft/pyright#2585]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
When inferring the type of a member assigned a Literal value passed as a parameter to a class initializer, the wrong type is inferred. Instead of correctly determining the type to be equivalent to the `Literal`, it infers a type equal to the union of the Literal's parameters.

**To Reproduce**
Type check the following code:

```python
from typing import Literal

SomeLiteral = Literal["a", 1]

class X:

    def __init__(self, y: SomeLiteral):
        self._y = y

    @property
    def y(self) -> SomeLiteral:
        return self._y # <-- INCORRECT ERROR: Expression of type "str | int" cannot be assigned to return type "SomeLiteral"
```

**Expected behavior**
The expected behaviour should be for Pyright to infer the narrowest type. Explictly stating the type as follows works as expected:

```python
from typing import Literal

SomeLiteral = Literal["a", 1]

class X:
    _y: SomeLiteral

    def __init__(self, y: SomeLiteral):
        self._y = y

    @property
    def y(self) -> SomeLiteral:
        return self._y
```

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/3920249/142682507-2c543216-2a4d-4c2c-bd49-a1d0b5902197.png)

**VS Code extension or command-line**
Both CLI and VS Code elicit the incorrect error.

**Comment by erictraut** [comment]
This behavior is intended. You haven't provided an explicit type annotation for the `_y` instance variable. Pyright therefore uses its standard type inference rules to infer the type. As part of those rules, it widens literals. This is consistent with the intent of most code that follows this pattern.

If your intent is for an instance variable to be constrained to a literal type, you should provide an explicit type annotation.

**Comment by SandyChapman** [comment]
Okay, thanks for the clarification. I don't find it intuitive for it to work this way, but if that's the way it is then that's the way it is.
